### ISSUE TITLE
Zero-shot protein function prediction performance is bad

### ISSUE THREAD

**Issue by nowittynamesleft** [nowittynamesleft/protein_function_description#21]
Background on function prediction with this model:

Currently we have a function prediction model that is capable of zero-shot prediction, or description "retrieval". It works in the following way:
The transformer encoder-decoder model is able to assign a probability for a given sequence set and natural language functional description. We can take all pairs, including mismatched ones, of sequence sets and GO term descriptions, and compute probabilities that the model assigns to the pairs. For a given sequence set, we take the top k descriptions in terms of assigned probability, and see if the actual GO term description for that sequence set is in the top k descriptions. We can compute top k accuracy in this way. 

Note: this works for descriptions that were not trained on, or even descriptions not part of GO at all; I could make up a description that is differently-worded but similar in meaning to a given GO term, and see if the model assigns a high probability to the proteins that are annotated with the given GO term.

With this way of function prediction, I can train the model on a set of sequences and GO term descriptions, and evaluate the model in a zero-shot setting on GO terms not included in the training set, where the classification is over all GO terms (including the training set GO terms)

The actual issue:
Currently, the top k performance when doing this is very low. (top 1000 accuracy is 27%, top 100 is 5.58%, top 1 is 0.09%).

This way of evaluation is ok, but probably could be improved. The sequence set that is input into the model can have many possible GO terms associated with them, as whatever GO term is corresponding to them has ancestor terms that may also be trained on the same subsampled sequence set. 
Two things we could do to improve this situation of difference in evaluation and training, that honestly might be in conflict with each other:
- We might change the evaluation to be inclusive of ALL possible correct descriptions of the sequence set by calculating AUPR. (Could add to #11  )
- We can instead define training to be "for a given sequence set, output the MOST SPECIFIC possible description for this set." Meaning, constrain the subsampling of any particular term's sequence set to have only that term as the possible answer, do not let the model subsample a sequence set that has a more specific term that could accurately be applied to the whole set.

There are several other things we can do to improve the performance of the model by improving the learned probability distribution:
Issues #17 (scheduled sampling) #15 (byte-pair encoding) and #20 (label smoothing) all have potential to improve training and therefore this function prediction performance. We need to systematically evaluate these additions, with their associated hyperparameters, on a validation set, and find the best possible set of hyperparameters.
